pwiz_tools/Shared/PanoramaClient/PanoramaFolderBrowser.cs

```diff
@@ -444,54 +444,40 @@ public static void AddChildContainers(TreeNode node, JToken folder, bool require
         var subFolders = folder[@"children"].Children();
         foreach (var subFolder in subFolders)
         {
+            if (!PanoramaUtil.CheckReadPermissions(subFolder))
+            {
+                // Do not add the folder if user does not have read permissions in the folder. 
+                // Any subfolders, even if they have read permissions, will also not be added.
+                continue;
+            }
+
             var folderName = (string)subFolder[@"name"];
 
             var folderNode = new TreeNode(folderName);
             AddChildContainers(folderNode, subFolder, requireUploadPerms, showFiles, server);
 
+            var isTargetedMsFolder = PanoramaUtil.IsTargetedMsFolder(subFolder);
             // User can only upload to folders where TargetedMS is an active module.
-            bool canUpload;
+            var canUpload = PanoramaUtil.CheckInsertPermissions(subFolder) && isTargetedMsFolder;
 
-            if (requireUploadPerms)
-            {
-                canUpload = PanoramaUtil.CheckFolderPermissions(subFolder) &&
-                            PanoramaUtil.CheckFolderType(subFolder);
-            }
-            else
-            {
-                var userPermissions = subFolder.Value<int?>(@"userPermissions");
-                canUpload = userPermissions != null && Equals(userPermissions & 1, 1);
-            }
-            // If the user does not have write permissions in this folder or any
-            // of its subfolders, do not add it to the tree.
-            if (requireUploadPerms)
-            {
-                if (folderNode.Nodes.Count == 0 && !canUpload)
-                {
-                    continue;
-                }
-            }
-            else
+            if (requireUploadPerms && folderNode.Nodes.Count == 0 && !canUpload)
             {
-                if (!canUpload)
-                {
-                    continue;
-                }
+                // If the user does not have write permissions in this folder or any
+                // of its subfolders, do not add it to the tree.
+                continue;
             }
 
-
-
             node.Nodes.Add(folderNode);
 
-            // User cannot upload files to folder
-            if (!canUpload)
+            if (requireUploadPerms && !(isTargetedMsFolder && canUpload))
             {
+                // User cannot upload files to folder
                 folderNode.ForeColor = Color.Gray;
                 folderNode.ImageIndex = folderNode.SelectedImageIndex = (int)ImageId.folder;
             }
             else
             {
-                if ((PanoramaUtil.CheckFolderType(subFolder) && !requireUploadPerms) || requireUploadPerms)
+                if (isTargetedMsFolder)
```

Comment:
Make else if, and maybe make it !isTargetedMsFolder to put shorter condition first.